### PR TITLE
guix: Use `gcc-8` across the board

### PR DESCRIPTION
**Pull request by dongcarl**
Only non-base commit is the last commit: https://github.com/bitcoin/bitcoin/commit/b5abb07d0d8051f337df2e7981d42a60e0140ee3

Right now, here's what we use in Gitian:
- Linux: Focal's [`g++-8-<arch>-linux-gnu`](https://packages.ubuntu.com/focal/g++-8-aarch64-linux-gnu) (`8.4.0-3ubuntu1cross1`)
- MinGW-w64: Focal's [`g++-mingw-w64`](https://packages.ubuntu.com/focal/g++-mingw-w64) (`9.3.0-7ubuntu1+22~exp1ubuntu4`)

In Guix right now we use `gcc-9` across the board.

I think it makes more sense to use `gcc-8` across the board, as it doesn't suffer from the `memcmp` bug, and is what debian buster (stable) does, meaning it will be well tested ([`g++-mingw-w64`](https://packages.debian.org/buster/g++-mingw-w64), [`g++-aarch64-linux-gnu`](https://packages.debian.org/buster/g++-aarch64-linux-gnu)).

We can accomplish this somewhat easily using Guix as we have tighter control over the toolchain (see: https://github.com/bitcoin/bitcoin/commit/b5abb07d0d8051f337df2e7981d42a60e0140ee3).

Let me know your thoughts!

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK on keeping the same compiler while migrating from gitian to guix.

**Comment by dongcarl** [comment]
> Concept ACK on keeping the same compiler while migrating from gitian to guix.

Note that we won't be keeping the same compiler for mingw-w64 (but I think that's okay) :-)

**Comment by MarcoFalke** [comment]
The switch from gitian->guix will be a major bump, so I think we don't need to use the exact same versions for all packages. However, looking at Debian (and Ubuntu) to see which packages are well-propagated among users (and thus tested) makes sense. I presume, once Debian/Ubuntu upgrade from gcc-10.2 to gcc-10.3, we could use that.

Concept ACK to prefer 8 over 9 in the meantime.

**Comment by sipa** [comment]
I think using GCC 8 is decent choice.

**Comment by dongcarl** [comment]
Pushed b5abb07d0d8051f337df2e7981d42a60e0140ee3 -> c90f6e51094a1ba4fb2aab35b78f23b6fda645d0
- Updated to be based on `master` instead of on #21462
- Updated commit description

**Comment by MarcoFalke** [comment]
Approach ACK c90f6e51094a1ba4fb2aab35b78f23b6fda645d0, haven't reviewed

File: contrib/guix/libexec/build.sh [hunk 1]
```diff
@@ -255,7 +255,7 @@ case "$HOST" in
 esac
 
 case "$HOST" in
-    powerpc64-linux-*) HOST_LDFLAGS="${HOST_LDFLAGS} -Wl,-z,noexecstack" ;;
+    powerpc64-linux-*|riscv64-linux-*) HOST_LDFLAGS="${HOST_LDFLAGS} -Wl,-z,noexecstack" ;;
```
Comment (hebasto) [inline contrib/guix/libexec/build.sh:258]: Why this change?

Comment (dongcarl) [reply contrib/guix/libexec/build.sh:258]: It was required to pass the `NX` check for `riscv64`, I'm guessing the default `execstack` behavior changed between gcc 8 and 9

Comment (dongcarl) [reply contrib/guix/libexec/build.sh:258]: Found it!
```diff
commit 12af29ab10fb5677cf947bcd7530c5f57c97522e
Author: Jim Wilson <jimw@sifive.com>
Date:   Sat Jul 14 20:05:39 2018 +0000

    RISC-V: Fix nested function trampolines.
    
            gcc/
            * config/riscv/linux.h (TARGET_ASM_FILE_END): New.
    
    From-SVN: r262660

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index 776a3c4434d..9179987e98c 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,7 @@
+2018-07-14  Jim Wilson  <jimw@sifive.com>
+
+	* config/riscv/linux.h (TARGET_ASM_FILE_END): New.
+
 2018-07-14  Paul Koning  <ni1d@arrl.net>
 
 	* config/pdp11/pdp11.c (pdp11_rtx_costs): Bugfixes.
diff --git a/gcc/config/riscv/linux.h b/gcc/config/riscv/linux.h
index 85561846dad..e208c95fe13 100644
--- a/gcc/config/riscv/linux.h
+++ b/gcc/config/riscv/linux.h
@@ -66,3 +66,5 @@ along with GCC; see the file COPYING3.  If not see
       %{rdynamic:-export-dynamic} \
       -dynamic-linker " GNU_USER_DYNAMIC_LINKER "} \
     %{static:-static}}"
+
+#define TARGET_ASM_FILE_END file_end_indicate_exec_stack
```

```
$ git tag --contains 12af29ab10fb5677cf947bcd7530c5f57c97522e
basepoints/gcc-10
basepoints/gcc-11
basepoints/gcc-12
misc/cutover-git
misc/first-auto-changelog
misc/first-auto-changelog-10
misc/first-auto-changelog-9
releases/gcc-10.1.0
releases/gcc-10.2.0
releases/gcc-10.3.0
releases/gcc-11.1.0
releases/gcc-9.1.0
releases/gcc-9.2.0
releases/gcc-9.3.0

**Review by hebasto — APPROVED** [review]
ACK c90f6e51094a1ba4fb2aab35b78f23b6fda645d0, I have reviewed the code and it looks OK, I agree it can be merged.

File: contrib/guix/libexec/build.sh (hunk 1 above)
Comment (laanwj) [reply contrib/guix/libexec/build.sh:258]: Correct, this used to be needed for RISC-V as well. We did have it in the gitian build. However, for the gitian build it was removed in 2ecaf214331b506ebfac4f4922241744357d652b.
(I don't understand why it's no longer needed on Focal's g++8, maybe Ubuntu is patching it somehow?)

Comment (fanquake) [reply contrib/guix/libexec/build.sh:258]: I've double-checked that when building RISC-V binaries using `g++-8-riscv64-linux-gnu` on Focal (20.04), the stack isn't being marked as executable. I did the following inside a new Ubuntu Focal Docker container:
```bash
# binutils for readelf
apt update && apt install g++-8-riscv64-linux-gnu binutils-riscv64-linux-gnu binutils

# riscv64-linux-gnu-gcc-8 --version
riscv64-linux-gnu-gcc-8 (Ubuntu 8.4.0-3ubuntu1) 8.4.0

# compile minimal binary
echo "int main() { return 0; }" > test.c
riscv64-linux-gnu-gcc-8 test.c -o test

# check stack
readelf -lW test|grep GNU_STACK
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x10


# recompile, tell linker to mark stack as executable
riscv64-linux-gnu-gcc-8 test.c -o test -Wl,-z,execstack

# recheck stack
readelf -lW test|grep GNU_STACK
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RWE 0x10
```

If you perform the same inside a Ubuntu Bionic (18.04) container:
```bash
# binutils for readelf
apt update && apt install g++-8-riscv64-linux-gnu binutils-riscv64-linux-gnu binutils

# riscv64-linux-gnu-gcc-8 --version
riscv64-linux-gnu-gcc-8 (Ubuntu 8.4.0-1ubuntu1~18.04) 8.4.0

# compile minimal binary
echo "int main() { return 0; }" > test.c
riscv64-linux-gnu-gcc-8 test.c -o test

# check stack
readelf -lW test|grep GNU_STACK
< no output >

# recompile, tell linker to not mark stack as executable
riscv64-linux-gnu-gcc-8 test.c -o test -Wl,-z,noexecstack

# recheck stack
readelf -lW test|grep GNU_STACK
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x10
```

So it's less that the stack is being marked as executable when building on Bionic, and more that the `GNU_STACK` section is just missing entirely, unless you explicitly tell the linker you want either a executable or non-executable stack. This makes sense when combined with the change linked by @dongcarl above (more context [here](https://patchwork.ozlabs.org/project/gcc/patch/20180714200633.27663-1-jimw@sifive.com/)), as the `GNU_STACK` section only started being added by default in GCC 9+. However it doesn't explain why the section is being added when using GCC 8 on Focal (and no additional linker flags). My assumption was that the change had been backported / applied by Debian/Ubuntu, however I can't seem to find evidence of that so far.

Comment (laanwj) [reply contrib/guix/libexec/build.sh:258]: Thanks for checking!

> So it's less that the stack is being marked as executable when building on Bionic, and more that the GNU_STACK section is just missing entirely

Right—not having a GNU_STACK section is effectively the same as having a GNU_STACK section with RWE, so, executable stack. Our security check takes this into account.

**Comment by laanwj** [comment]
No strong opinion on what gcc to use, but I think the argument is fair that going with what we know is the safest choice. The change looks consistent, in any case.
Code review ACK c90f6e51094a1ba4fb2aab35b78f23b6fda645d0
